README
Radio Kill Switch
------------ -----   -----       ----       ---       --         -
Most laptops provide the ability for the user to physically disable the radio.
Some vendors have implemented this as a physical switch that requires no
software to turn the radio off and on.  On other laptops, however, the switch
is controlled through a button being pressed and a software driver then making
calls to turn the radio off and on.  This is referred to as a "software based
RF kill switch"

Currently this project provides modules for controlling the software RF kill
switch on the Averatec 5100P and Packard Bell EasyNote E5.  The code may work
on other laptops, but these are the only models on which it has been tested.

To determine if you have a system that might be compatible with one of the
provided SW RF Kill switch modules, you can run:


        To check for the Packard Bell (to use module pbe) --

        % dd if=/dev/mem bs=1 skip=983040 count=65535 2>/dev/null | strings | egrep "NEW-PC|Insyde Software MobilePRO BIOS"

        To check for the Averatec (to use module av5100) --

        % dd if=/dev/mem bs=1 skip=983040 count=65535 2>/dev/null | strings | egrep "AVERATEC"

If you have one of those laptop models you can imply loading the av5100/pbe5
module and the radio will be toggled on and off.  In addition, you can turn
the driver on and off by writing either a 1 or 0 to /proc/av5100/radio or
/proc/pbe5/radio.  If you automatically load the av5100/pbe5 module when your
system boots, you may wish to use the radio module parameter to control the
state of the radio upon loading:

        modprobe av5100 radio=0
        modprobe pbe5 radio=0

results in the module loading with the radio turned off.  You can then turn the
radio on by:

        echo 1 > /proc/av5100/radio
        echo 1 > /proc/pbe5/radio

If you have a SW RF kill switch and can not use one of the above modules,
please join us on IRC (irc.freenode.net) on channel #ipw2100 and someone may
be able to help.


debian/changelog
rfswitch (1.3-6) ORPHANED; urgency=low

  * NOT RELEASED YET
  * ORPHAN PACKAGE

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu, 30 Apr 2009 21:20:23 +0200

rfswitch (1.3-5) unstable; urgency=low

  * use sf-redirector in watch files, to make lintian happier.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 27 Jan 2009 17:40:14 +0100

rfswitch (1.3-4) unstable; urgency=low

  * switch from dh_clean -k to dh_prep, to become compliant with debhelper 7
    and to make lintian happy.
  * drop support for kernel 2.4.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Tue, 27 Jan 2009 17:37:36 +0100

rfswitch (1.3-3) unstable; urgency=low

  * make lintian happy and depend on ${misc:Depends}.
  * switch to dh7.
  * fix syntax error.
  * refresh patches.
  * add patch descriptions.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Mon, 08 Dec 2008 23:05:35 +0100

rfswitch (1.3-2) unstable; urgency=low

  * Convert to debhelper + quilt. Bump debhelper dependency to >= 7.
  * Add 01_2.6.26_compat.patch to fix compile against Linux 2.6.26.
  * Add 02_makefile_clean.patch to cleanup new kbuild by-products.

 -- Kel Modderman <kel@otaku42.de>  Tue, 15 Jul 2008 22:19:19 +1000

rfswitch (1.3-1) unstable; urgency=low

  * New upstream release.
  * drop patches, applied upstream.
  * drop dpatch, no longer necessary.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 05 Apr 2008 00:08:16 +0200

rfswitch (1.1-8) unstable; urgency=low

  * appease lintian and write (C) as © in debian/copyright.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Wed, 20 Feb 2008 23:47:43 +0100

rfswitch (1.1-7) unstable; urgency=low

  * update to debhelper compat 6.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 19 Jan 2008 01:03:29 +0100

rfswitch (1.1-6) unstable; urgency=low

  * use official Homepage tag.
  * add debian/watch.
  * update to Standards-Version 3.7.3, no changes necessary.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Sat, 08 Dec 2007 13:29:31 +0100

rfswitch (1.1-5) unstable; urgency=low

  * add VCS control headers.
  * update syntax for Vcs headers in debian/control

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu, 25 Oct 2007 19:03:04 +0200

rfswitch (1.1-4) unstable; urgency=low

  * Remove reference to obsolete kernel-image-_KVERS_ in
    debian/control.modules.in.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Thu, 02 Aug 2007 23:12:48 +0200

rfswitch (1.1-3) unstable; urgency=low

  * fix maintainer mail address.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Fri,  2 Mar 2007 03:22:35 +0100

rfswitch (1.1-2) unstable; urgency=low

  * fix compilation against kernel 2.6.19.

 -- Stefan Lippers-Hollmann <s.l-h@gmx.de>  Mon, 13 Nov 2006 19:12:41 +0100

rfswitch (1.1-1) unstable; urgency=low

  * New upstream release
  * Fix build-deps
  * Versioned depends on debhelper 5.0.37 for dh_installmodules
  * Remove patch and patch infrastructure

 -- Kel Modderman <kelmo@kanotixguide.org>  Wed,  4 Oct 2006 00:18:04 +1000

rfswitch (0.1-11) unstable; urgency=low

  * Add patch to correct bad Makefile behaviour with linux 2.6
  * Bump Standards-version to 3.7.2
  * Tweak module packing

 -- Kel Modderman <kelrin@tpg.com.au>  Sat, 27 May 2006 21:30:33 +1000

rfswitch (0.1-10) unstable; urgency=low

  * Bump debhelper compat level to 5
  * Add homepage to long description
  * Binary modules depend on linux-image
  * Binary modules provide rfswitch-modules
  * Remove bogus targets from rules.modules
  * Use dh_installmodules to handle modules post{inst,rm}

 -- Kel Modderman <kelrin@tpg.com.au>  Mon, 20 Mar 2006 23:36:21 +1000

rfswitch (0.1-9) unstable; urgency=low

  * Clean up debian/rules.
    - loop install debian files
    - use bzip2
  * bzip2 added to relevant dependencies.
  * build-essential removed from build-dep field.
  * Remove cdbs auto update of control file.
    - removed control.in.
  * Add README.debian.

 -- Kel Modderman <kelrin@tpg.com.au>  Fri, 30 Sep 2005 03:41:46 +1000

rfswitch (0.1-8) unstable; urgency=low

  * Remove debconf depends from postinst.
  * Adjust copyright file.

 -- Kel Modderman <kelrin@tpg.com.au>  Fri, 02 Sep 2005 23:22:57 +1000

rfswitch (0.1-7) unstable; urgency=low

  * Removed useless code from rules.modules.
  * Removed depends on kernel-package.
  * Registered with mod-ass front-end.

 -- Kel Modderman <kelrin@tpg.com.au>  Fri, 26 Aug 2005 21:35:18 +1000

rfswitch (0.1-6) unstable; urgency=low

  * Changed name of source package.

 -- Kel Modderman <kelrin@tpg.com.au>  Thu, 18 Aug 2005 23:30:16 +1000

rfswitch-source (0.1-5) unstable; urgency=low

  * cdbs now handles rfswitch-source preparation.
    - added rfswitch-source.install
    - mod-ass targets moved to rules.modules

 -- Kel Modderman <kelrin@tpg.com.au>  Wed, 17 Aug 2005 12:24:36 +1000

rfswitch-source (0.1-4) unstable; urgency=low

  * Improved kernel build command.
  * Cleaned up kdist_clean target.
  * Removed nostrip option.

 -- Kel Modderman <kelrin@tpg.com.au>  Thu, 11 Aug 2005 13:55:08 +1000

rfswitch-source (0.1-3) unstable; urgency=low

  * Added CFLAGS / DEB_BUILD_OPTIONS options.
  * Moved build-indep rules to install-indep.

 -- Kel Modderman <kelrin@tpg.com.au>  Fri, 05 Aug 2005 13:36:54 +1000

rfswitch-source (0.1-2) unstable; urgency=low

  * Revamped dehelper and kernel-package targets.
  * Defined PACKAGE.
  * Added modextension variable for kernel module extensions.
  * Added rfswitch-source.docs.

 -- Kel Modderman <kelrin@tpg.com.au>  Sun, 31 Jul 2005 02:02:26 +1000

rfswitch-source (0.1-1) unstable; urgency=low

  * Initial release.

 -- Kel Modderman <kelrin@tpg.com.au>  Thu,  7 Jul 2005 23:44:17 +1000



av5100.c
/*******************************************************************************
  
  Copyright(c) 2003 - 2004 Intel Corporation. All rights reserved.
  
  This program is free software; you can redistribute it and/or modify it 
  under the terms of version 2 of the GNU General Public License as 
  published by the Free Software Foundation.
  
  This program is distributed in the hope that it will be useful, but WITHOUT 
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for 
  more details.
  
  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 59 
  Temple Place - Suite 330, Boston, MA  02111-1307, USA.
  
  The full GNU General Public License is included in this distribution in the
  file called LICENSE.
  
  Contact Information:
  James P. Ketrenos <ipw2100-admin@linux.intel.com>
  Intel Corporation, 5200 N.E. Elam Young Parkway, Hillsboro, OR 97124-6497

*******************************************************************************/
#include <linux/compiler.h>
#include <linux/errno.h>
#include <linux/kernel.h>
#include <linux/module.h>
#include <linux/types.h>
#include <linux/netdevice.h>
#include <linux/version.h>
#include <linux/proc_fs.h>
#include <linux/ioport.h>
#include <asm/uaccess.h>
#include <asm/io.h>


#define DRV_NAME		"av5100"
#define DRV_VERSION		"1.3"
#define DRV_DESCRIPTION		"SW RF kill switch for Averatec 5100P"
#define DRV_COPYRIGHT		"Copyright(c) 2003-2004 Intel Corporation"

static int radio = 1;

#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0)

MODULE_PARM(radio, "i");

#else /* LINUX_VERSION_CODE < 2.6.0 */

#include <linux/moduleparam.h>
module_param(radio, int, 1);

#endif /* LINUX_VERSION_CODE < 2.6.0 */

MODULE_PARM_DESC(radio, "controls state of radio (1=on, 0=off)");

MODULE_DESCRIPTION(DRV_DESCRIPTION);
MODULE_AUTHOR(DRV_COPYRIGHT);
MODULE_LICENSE("GPL");

#define AV5100_RADIO_ON (0xe0)
#define AV5100_RADIO_OFF (0xe1)

static int av5100_radio = AV5100_RADIO_OFF;

static void av5100_set_radio(int state)
{
	printk(KERN_INFO DRV_NAME ": Radio being turned %s\n",
	       (state  == AV5100_RADIO_ON) ? "ON" : "OFF");
	outl(0x80020800, 0xcf8);
	outb(0x6f, 0x0072);
	outl(0x1800ffff, 0x1184); 
	outb(state, 0x00b2);
	av5100_radio = state;
}


/*
 * proc stuff
 */
static struct proc_dir_entry *dir_base = NULL;

static int proc_set_radio(struct file *file, const char *buffer, 
			  unsigned long count, void *data)
{
	av5100_set_radio(buffer[0] == '0' ? AV5100_RADIO_OFF : AV5100_RADIO_ON);
	
	return count;
}

static int proc_get_radio(char *page, char **start, off_t offset,
			  int count, int *eof, void *data)
{
	int len = 0;
	
	len += snprintf(page, count, DRV_NAME ": %d\n", 
			av5100_radio == AV5100_RADIO_OFF ? 0 : 1);
	
	*eof = 1;
	return len;
}


static void av5100_proc_cleanup(void)
{
	if (dir_base) {
		remove_proc_entry("radio", dir_base);
		remove_proc_entry(DRV_NAME, &proc_root);
		dir_base = NULL;
	}
}


static int av5100_proc_init(void)
{
	struct proc_dir_entry *ent;
	int err = 0;

	dir_base = create_proc_entry(DRV_NAME, S_IFDIR, &proc_root);
	if (dir_base == NULL) {
		printk(KERN_ERR DRV_NAME ": Unable to initialise /proc/" 
		       DRV_NAME "\n");
		err = -ENOMEM;
		goto fail;
	}


	ent = create_proc_entry("radio", S_IFREG | S_IRUGO | S_IWUSR, 
				dir_base);
	if (ent) {
		ent->read_proc = proc_get_radio;
		ent->write_proc = proc_set_radio;
	} else {
		printk(KERN_ERR
		       "Unable to initialize /proc/" DRV_NAME "/radio\n");
		err = -ENOMEM;
		goto fail;
	}

	return 0;

 fail:
	av5100_proc_cleanup();
	return err;
}

/*
 * module stuff
 */
static int __init av5100_init(void)
{
	av5100_proc_init();

	av5100_set_radio((radio == 1) ? AV5100_RADIO_ON : AV5100_RADIO_OFF);
	
	return 0;
}

static void __exit av5100_exit(void)
{
	av5100_set_radio(AV5100_RADIO_OFF);

	av5100_proc_cleanup();
}

module_init(av5100_init);
module_exit(av5100_exit);

/*
         1         2         3         4         5         6         7
12345678901234567890123456789012345678901234567890123456789012345678901234567890
*/


pbe5.c
/*******************************************************************************
  
  This program is free software; you can redistribute it and/or modify it 
  under the terms of version 2 of the GNU General Public License as 
  published by the Free Software Foundation.
  
  This program is distributed in the hope that it will be useful, but WITHOUT 
  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for 
  more details.
  
  You should have received a copy of the GNU General Public License along with
  this program; if not, write to the Free Software Foundation, Inc., 59 
  Temple Place - Suite 330, Boston, MA  02111-1307, USA.
  
  The full GNU General Public License is included in this distribution in the
  file called LICENSE.
  
  Author:
  Pedro Ramalhais <pmr09313@students.fct.unl.pt>
  
  Based on:
  av5100.c from http://ipw2100.sourceforge.net/

*******************************************************************************/

#include <linux/compiler.h>
#include <linux/errno.h>
#include <linux/kernel.h>
#include <linux/module.h>
#include <linux/types.h>
#include <linux/netdevice.h>
#include <linux/version.h>
#include <linux/proc_fs.h>
#include <linux/ioport.h>
#include <asm/uaccess.h>
#include <asm/io.h>

#define DRV_NAME		"pbe5"
#define DRV_VERSION		"1.3"
#define DRV_DESCRIPTION		"SW RF kill switch for Packard Bell EasyNote E5"
#define DRV_AUTHOR		"Pedro Ramalhais"
#define DRV_LICENSE		"GPL"

static int radio = 1;

#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0)

MODULE_PARM(radio, "i");

#else /* LINUX_VERSION_CODE < 2.6.0 */

#include <linux/moduleparam.h>
module_param(radio, int, 1);

#endif /* LINUX_VERSION_CODE < 2.6.0 */

MODULE_PARM_DESC(radio, "controls state of radio (1=on, 0=off)");

MODULE_DESCRIPTION(DRV_DESCRIPTION);
MODULE_AUTHOR(DRV_AUTHOR);
MODULE_LICENSE(DRV_LICENSE);

/*
 * NOTE: These values were obtained from disassembling the Icon.exe program
 * installed in the Packard Bell EasyNote E5 laptop. The names were guessed,
 * so don't rely on them.
 */
#define PBE5_PORT_TOGGLE	0x0b3
#define PBE5_VALUE_TOGGLE_ON	0x01
#define PBE5_VALUE_TOGGLE_OFF	0x00
#define PBE5_PORT_APPLY		0x0b2
#define PBE5_VALUE_APPLY	0xef

// Some "booleans" =;-)
#define PBE5_RADIO_OFF	0
#define PBE5_RADIO_ON	1

static int pbe5_radio_status = PBE5_RADIO_ON;

unsigned char pbe5_get_radio(void)
{
	unsigned char val = 0x00;
	
	val = inb(PBE5_PORT_TOGGLE);
	
	return val;
}

static void pbe5_set_radio(int state_set)
{
	pbe5_radio_status = pbe5_get_radio();

	if (pbe5_radio_status != state_set) {
		// Set the radio toggle register
		outb(PBE5_VALUE_TOGGLE_ON, PBE5_PORT_TOGGLE);
		// Commit the radio toggle register value
		outb(PBE5_VALUE_APPLY, PBE5_PORT_APPLY);
		// Update the radio status
		pbe5_radio_status = pbe5_get_radio();

		printk(KERN_INFO DRV_NAME ": Radio turned %s\n",
			(state_set  == PBE5_RADIO_ON) ? "ON" : "OFF");
	} else {
		printk(KERN_INFO DRV_NAME ": Radio already %s\n",
			(state_set  == PBE5_RADIO_ON) ? "ON" : "OFF");
	}
}


/*
 * proc stuff
 */
static struct proc_dir_entry *dir_base = NULL;

static int proc_set_radio(struct file *file, const char *buffer, 
			  unsigned long count, void *data)
{
	pbe5_set_radio(buffer[0] == '0' ? PBE5_RADIO_OFF : PBE5_RADIO_ON);
	
	return count;
}

static int proc_get_radio(char *page, char **start, off_t offset,
			  int count, int *eof, void *data)
{
	int len = 0;
	
	len += snprintf(page, count, DRV_NAME ": %d\n", 
			pbe5_radio_status == PBE5_RADIO_OFF ? 0 : 1);
	
	*eof = 1;
	return len;
}


static void pbe5_proc_cleanup(void)
{
	if (dir_base) {
		remove_proc_entry("radio", dir_base);
		remove_proc_entry(DRV_NAME, &proc_root);
		dir_base = NULL;
	}
}


static int pbe5_proc_init(void)
{
	struct proc_dir_entry *ent;
	int err = 0;

	dir_base = create_proc_entry(DRV_NAME, S_IFDIR, &proc_root);
	if (dir_base == NULL) {
		printk(KERN_ERR DRV_NAME ": Unable to initialise /proc/" 
		       DRV_NAME "\n");
		err = -ENOMEM;
		goto fail;
	}


	ent = create_proc_entry("radio", S_IFREG | S_IRUGO | S_IWUSR, 
				dir_base);
	if (ent) {
		ent->read_proc = proc_get_radio;
		ent->write_proc = proc_set_radio;
	} else {
		printk(KERN_ERR
		       "Unable to initialize /proc/" DRV_NAME "/radio\n");
		err = -ENOMEM;
		goto fail;
	}

	return 0;

 fail:
	pbe5_proc_cleanup();
	return err;
}

/*
 * module stuff
 */
static int __init pbe5_init(void)
{
	pbe5_proc_init();

	pbe5_set_radio((radio == 1) ? PBE5_RADIO_ON : PBE5_RADIO_OFF);
	
	return 0;
}

static void __exit pbe5_exit(void)
{
	pbe5_set_radio(PBE5_RADIO_OFF);

	pbe5_proc_cleanup();
}

module_init(pbe5_init);
module_exit(pbe5_exit);

/*
         1         2         3         4         5         6         7
12345678901234567890123456789012345678901234567890123456789012345678901234567890
*/
